### PR TITLE
Bug 1680462 - Dont show search suggestions when disabled

### PR DESCRIPTION
**Pull request by daleharvey**

### CODE REVIEW

File: tests/browser/browser_test.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ const ABOUT_BLANK = "about:blank";
 const URLBAR_PROVIDER_NAME = "ProviderDynamicQuickSuggest";
 
 const ATTRIBUTION_URL =
-  "http://mochi.test:8888/browser/testing/extensions/browser/qs_attribution.sjs";
+  "http://mochi.test:8888/browser/testing/extensions/browser/qs_attribution.sjs?";
```
Comment (daleharvey) [inline tests/browser/browser_test.js:24]: This needed added due to a change in PartnerLinkAttribution that appends the campaignID expecting that there is already a queryString

**Comment by daleharvey** [comment]
Ideally would add tests for private browsing mode as well, however these tests until they run on CI are quite brittle so I would rather have a few working tests than lots of broken ones, can add them if you think they are needed though.

**Review by htwyford — CHANGES_REQUESTED** [review]
I don't think this check being a part of `matchSearchTerm` makes sense. I'd prefer if you created a new experimental API – maybe `suggestionsEnabled` – that returns a boolean. Then you can call it from `isActive`. I could see `suggestionsEnabled` being useful to other extensions, which would be a plus.

Maybe have `suggestionsEnabled` take a boolean, `isPrivate`, and base the returned value on that rather than getting the top window. So something like

```
// src/background.js -> isActive
let suggestionsEnabled = await browser.experiments.urlbar.suggestionsEnabled(queryContext.isPrivate);
if (!suggestionsEnabled) {
  return;
}

// api.js
XPCOMUtils.defineLazyPreferenceGetter(
  this,
  "searchSuggestionsEnabled",
  "browser.search.suggest.enabled",
  true
);

XPCOMUtils.defineLazyPreferenceGetter(
  this,
  "privateSearchSuggestionsEnabled",
  "browser.search.suggest.enabled.private",
  false
);


...


async function suggestionsEnabled(isPrivate) {
  if (isPrivate) {
    return privateSearchSuggestionsEnabled;
  }
  return searchSuggestionsEnabled;
}

**Comment by daleharvey** [comment]
Cheers Harry, yeh using the BrowserWindowTracker seemed less than ideal so I do like this much better

File: src/background.js [hunk 2]
```diff
@@ -36,6 +36,9 @@ class ProviderDynamicQuickSuggest extends UrlbarProvider {
   }
 
   async isActive(queryContext) {
+    if (!(await api.suggestionsEnabled(queryContext.isPrivate))) {
+      return false;
+    }
     this.matchedResult = await api.matchSearchTerm(queryContext.searchString);
```
Comment (htwyford) [inline src/background.js]: After calling async code in `isActive` or `startQuery`, you'll want to check that `isActive`/`startQuery` haven't been called again -- i.e., the user typed another character in their query. There's boilerplate to do this. At the beginning of the function, we [create an empty object and assign it to a class variable](https://github.com/mozilla-extensions/firefox-quick-suggest-weather/blob/master/src/background.js#L226-L227). Then, after async calls, we [check that the two are still the same](https://github.com/mozilla-extensions/firefox-quick-suggest-weather/blob/master/src/background.js#L262-L264). Could you please add that after these two calls?

**Review by htwyford — APPROVED** [review]
Looks good, thanks! Just one more change

File: src/background.js [hunk 3]
```diff
@@ -36,6 +36,14 @@ class ProviderDynamicQuickSuggest extends UrlbarProvider {
...
 
   async isActive(queryContext) {
+    let queryInstance = {};
+    this.queryInstance = queryInstance;
+    if (!(await api.suggestionsEnabled(queryContext.isPrivate))) {
+      return false;
+    }
+    if (this.queryInstance != queryInstance) {
+      return false;
+    }
```
Comment (htwyford) [inline src/background.js:47]: Please also put this below `await api.matchSearchTerm`

File: tests/browser/browser_test.js [hunk 4]
```diff
@@ -116,6 +121,27 @@ add_task(async function test_attribution() {
   SpecialPowers.setCharPref("browser.partnerlink.attributionURL", "");
 });
 
+add_task(async function test_suggestions_disabled() {
```
Comment (htwyford) [inline tests/browser/browser_test.js:127]: Please also add a test for private windows

**Review by htwyford — APPROVED** [review]

**Pull request merged**